### PR TITLE
add reusable for win remote

### PR DESCRIPTION
**Pull request by acured**

### CODE REVIEW

File: ts/nni_manager/training_service/reusable/trialDispatcher.ts [hunk 1]
```diff
@@ -31,6 +32,8 @@ import { SharedStorageService, SharedStorageConfig } from './sharedStorage';
 import { NFSSharedStorageService } from './shared_storages/nfsStorageService'
 import { AzureBlobSharedStorageService } from './shared_storages/azureblobStorageService'
 import { TrialDetail } from './trial';
+import { LinuxCommands } from "../remote_machine/extends/linuxCommands";
```
Comment (SparkSnail) [inline ts/nni_manager/training_service/reusable/trialDispatcher.ts:35]: trialDispatcher.ts is a common service, should not import common commands from remote_machine folder.

File: ts/nni_manager/training_service/reusable/trialDispatcher.ts [hunk 2]
```diff
@@ -677,14 +682,8 @@ class TrialDispatcher implements TrainingService {
...
-        environment.command = `sh ../install_nni.sh && python3 -m nni.tools.trial_tool.trial_runner`;
-
-        if (this.isDeveloping) {
-            environment.command = "[ -d \"nni_trial_tool\" ] && echo \"nni_trial_tool exists already\" || (mkdir ./nni_trial_tool && tar -xof ../nni_trial_tool.tar.gz -C ./nni_trial_tool) && pip3 install websockets && " + environment.command;
-        }
-
-        environment.command = `mkdir -p envs/${envId} && cd envs/${envId} && ${environment.command}`;
-
+        environment.command = new LinuxCommands().reusableStartCommand(envId, this.isDeveloping);
+        environment.commandforwin = new WindowsCommands().reusableStartCommand(envId, this.isDeveloping);
```
Comment (SparkSnail) [inline ts/nni_manager/training_service/reusable/trialDispatcher.ts:686]: This `commandforwin` seems only for remote windows mode in this PR, don't put it here, this trialDispatcher is for common commands. Suggest to refer localEnvironmentService to see how it process local windows commands, or refactor logic to extract windows command and put it here.

File: ts/nni_manager/training_service/reusable/trialDispatcher.ts (hunk 1 above)
Comment (acured) [reply ts/nni_manager/training_service/reusable/trialDispatcher.ts]: Removed

File: ts/nni_manager/training_service/reusable/trialDispatcher.ts (hunk 2 above)
Comment (acured) [reply ts/nni_manager/training_service/reusable/trialDispatcher.ts]: Fixed.

File: ts/nni_manager/training_service/reusable/trialDispatcher.ts [hunk 3]
```diff
@@ -677,14 +679,7 @@ class TrialDispatcher implements TrainingService {
...
         this.log.info(`Assign environment service ${environmentService.getName} to environment ${envId}`);
-        environment.command = `sh ../install_nni.sh && python3 -m nni.tools.trial_tool.trial_runner`;
-
-        if (this.isDeveloping) {
-            environment.command = "[ -d \"nni_trial_tool\" ] && echo \"nni_trial_tool exists already\" || (mkdir ./nni_trial_tool && tar -xof ../nni_trial_tool.tar.gz -C ./nni_trial_tool) && pip3 install websockets && " + environment.command;
-        }
-
-        environment.command = `mkdir -p envs/${envId} && cd envs/${envId} && ${environment.command}`;
-
+        environment.command = `python3 -m nni.tools.trial_tool.trial_runner`;
```
Comment (SparkSnail) [inline ts/nni_manager/training_service/reusable/trialDispatcher.ts:682]: This change will break other training service.

Comment (acured) [reply ts/nni_manager/training_service/reusable/trialDispatcher.ts]: Reverted.

File: ts/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts [hunk 4]
```diff
@@ -208,6 +208,33 @@ export class RemoteEnvironmentService extends EnvironmentService {
...
     }
 
+    private async getScript(environment: EnvironmentInformation): Promise<string> {
+        const executor = await this.getExecutor(environment.id);
+        const logLevel = getLogLevel();
+        let script: string = environment.command;
+
+        if (executor.getIsWindows()) {
+            const prepare = `envs\\install_nni.cmd && mkdir envs\\${environment.envId} && cd envs\\${environment.envId}`;
+            const startrun = `cd .. && install_nni.cmd && cd ${environment.envId} && python -m nni.tools.trial_tool.trial_runner`;
```
Comment (J-shang) [inline ts/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts:218]: seems install twice? `envs\\install_nni.cmd` and `install_nni.cmd`

and multi `cd` can be simplified.

File: ts/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts [hunk 5]
```diff
@@ -208,6 +208,33 @@ export class RemoteEnvironmentService extends EnvironmentService {
...
+            const developingScript = "IF EXIST nni_trial_tool (ECHO \"nni_trial_tool exists already\") ELSE (mkdir nni_trial_tool && tar -xof ../nni_trial_tool.tar.gz -C ./nni_trial_tool) && pip3 install websockets";
+            
+            script = logLevel == "debug" ? `${prepare} && ${developingScript} && ${startrun}` : `${prepare} && ${startrun}`;
+        } else {
+            const prepare = `mkdir -p envs/${environment.envId} && cd envs/${environment.envId}`;
+            const startrun = `sh ../install_nni.sh && python3 -m nni.tools.trial_tool.trial_runner`;
+            const developingScript = "[ -d \"nni_trial_tool\" ] && echo \"nni_trial_tool exists already\" || (mkdir ./nni_trial_tool && tar -xof ../nni_trial_tool.tar.gz -C ./nni_trial_tool) && pip3 install websockets";
```
Comment (SparkSnail) [inline ts/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts:225]: Seems these linux commands already set in trialDispatcher.ts and could use original commands directly, only need to add windows comands in remoteEnvironmentService.ts?

File: ts/nni_manager/training_service/remote_machine/shellExecutor.ts [hunk 6]
```diff
@@ -110,6 +109,15 @@ class ShellExecutor {
         this.sshClient.end();
     }
 
+    public async getIsWindows(): Promise<boolean>{
+        const result = await this.execute("ver")
```
Comment (J-shang) [inline ts/nni_manager/training_service/remote_machine/shellExecutor.ts:113]: We have `ShellExecutor.isWindows` to check if is windows, if necessary, change `isWindows` to `public`.

File: ts/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts (hunk 5 above)
Comment (acured) [reply ts/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts]: Right, Fixed.

File: ts/nni_manager/training_service/reusable/trialDispatcher.ts [hunk 7]
```diff
@@ -225,8 +225,10 @@ class TrialDispatcher implements TrainingService {
             const codeFileName = await storageService.copyDirectory(codeDir, envDir, true);
             storageService.rename(codeFileName, "nni-code.tar.gz");
 
-            const installFileName = storageService.joinPath(envDir, 'install_nni.sh');
+            const installFileName = storageService.joinPath(envDir, `install_nni.sh`);
+            const installFileNameForWin = storageService.joinPath(envDir, `install_nni.cmd`);
```
[the remark this answers is not in the document]
Comment (acured) [reply ts/nni_manager/training_service/reusable/trialDispatcher.ts]: Done.

File: ts/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts (hunk 4 above)
Comment (acured) [reply ts/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts]: Fixed.

File: ts/nni_manager/training_service/remote_machine/shellExecutor.ts (hunk 6 above)
Comment (acured) [reply ts/nni_manager/training_service/remote_machine/shellExecutor.ts]: Done.

File: ts/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts [hunk 8]
```diff
@@ -208,6 +208,27 @@ export class RemoteEnvironmentService extends EnvironmentService {
...
+
+        if (executor.isWindows) {
+            const prepare = `mkdir envs\\${environment.id} && cd envs\\${environment.id}`;
+            const startrun = `powershell ..\\install_nni.ps1 && python -m nni.tools.trial_tool.trial_runner`;
+            const developingScript = "IF EXIST nni_trial_tool (ECHO \"nni_trial_tool exists already\") ELSE (mkdir nni_trial_tool && tar -xof ../nni_trial_tool.tar.gz -C ./nni_trial_tool) && pip3 install websockets";
+            
+            script = logLevel == "debug" ? `${prepare} && ${developingScript} && ${startrun}` : `${prepare} && ${startrun}`;
+        }
+
+        script = `cd ${environment.runnerWorkingFolder} && \
```
Comment (SparkSnail) [inline ts/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts:224]: this script value will override line 221.

**Review by J-shang — APPROVED** [review]

File: ts/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts (hunk 8 above)
Comment (acured) [reply ts/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts]: It only works for windows "isdebug" mode or not, Default(Linux) is not affected.

File: ts/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts [hunk 9]
```diff
@@ -291,7 +311,7 @@ export class RemoteEnvironmentService extends EnvironmentService {
         // Copy files in codeDir to remote working directory
         await executor.copyDirectoryToRemote(environmentLocalTempFolder, this.remoteExperimentRootDir);
         // Execute command in remote machine, set isInteractive=true to run script in conda environment
-        executor.executeScript(executor.joinPath(environment.runnerWorkingFolder,
+        executor.executeScript(executor.joinPath(this.remoteExperimentRootDir,
```
Comment (acured) [inline ts/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts:314]: @SparkSnail Can you take a look this change, I found it cause a issue when I rebase this branch.

File: ts/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts (hunk 8 above)
Comment (acured) [reply ts/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts]: FIxed.

**Review by SparkSnail — APPROVED** [review]

**Pull request merged**